### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by wangdayong228**
1. Return pointer when return value maybe nil.
2. Fix rpc metrics varaiable initialization
3. Refactor

---
This change is [Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45)

### CODE REVIEW

File: interface.go [hunk 1]
```diff
@@ -87,14 +86,14 @@ type ClientOperator interface {
 	GetAccountInfo(account types.Address, epoch ...*types.Epoch) (accountInfo types.AccountInfo, err error)
 	GetInterestRate(epoch ...*types.Epoch) (intersetRate *hexutil.Big, err error)
 	GetAccumulateInterestRate(epoch ...*types.Epoch) (intersetRate *hexutil.Big, err error)
-	GetBlockRewardInfo(epoch ...*types.Epoch) (rewardInfo []types.RewardInfo, err error)
+	GetBlockRewardInfo(epoch types.Epoch) (rewardInfo []types.RewardInfo, err error)
```
Comment (boqiu) [inline interface.go:89]: why not use point type, since the default epochs are all defined as pointer types.

File: rpc/metrics.go [hunk 2]
```diff
@@ -23,9 +23,9 @@ import (
 )
 
 var (
-	rpcRequestGauge        = metrics.NewRegisteredGauge("rpc/requests", nil)
-	successfulRequestGauge = metrics.NewRegisteredGauge("rpc/success", nil)
-	failedReqeustGauge     = metrics.NewRegisteredGauge("rpc/failure", nil)
+	rpcRequestGauge        = metrics.GetOrRegisterGauge("rpc/requests", nil)
+	successfulRequestGauge = metrics.GetOrRegisterGauge("rpc/success", nil)
+	failedReqeustGauge     = metrics.GetOrRegisterGauge("rpc/failure", nil)
 	rpcServingTimer        = metrics.NewRegisteredTimer("rpc/duration/all", nil)
```
Comment (boqiu) [inline rpc/metrics.go:29]: Why the last one not changed?

File: rpc/metrics.go [hunk 3]
```diff
@@ -23,9 +23,9 @@ import (
 )
 
 var (
-	rpcRequestGauge        = metrics.NewRegisteredGauge("rpc/requests", nil)
-	successfulRequestGauge = metrics.NewRegisteredGauge("rpc/success", nil)
-	failedReqeustGauge     = metrics.NewRegisteredGauge("rpc/failure", nil)
+	rpcRequestGauge        = metrics.GetOrRegisterGauge("rpc/requests", nil)
```
Comment (boqiu) [inline rpc/metrics.go:26]: moreover, if go-eth/rpc package also imported, better to ensure go-eth/rpc initialized at first.

So, you can explicitly import eth/rpc package in this file to ensure metrics are initialized correctly.

File: types/epoch.go [hunk 4]
```diff
@@ -48,11 +47,24 @@ func (e *Epoch) String() string {
 		return e.name
 	}
 
-	return hexutil.EncodeBig(e.number)
+	return e.number.String()
+}
+
+// IsEqual check if e equals input
+func (e *Epoch) IsEqual(input *Epoch) bool {
```
Comment (boqiu) [inline types/epoch.go:54]: In golang naming convention, `Equals` is a better name.

Also suggest to rename parameter `input` to `other`

File: types/epoch.go [hunk 5]
```diff
@@ -48,11 +47,24 @@ func (e *Epoch) String() string {
...
+	if e == nil && input == nil {
```
Comment (boqiu) [inline types/epoch.go:55]: `e` should not be nil

File: types/epoch.go [hunk 6]
```diff
@@ -48,11 +47,24 @@ func (e *Epoch) String() string {
...
+		return true
+	}
+	if e != nil || input != nil {
+		return false
+	}
+	isEqual := e.number.String() == input.number.String() && e.name == input.name
```
Comment (boqiu) [inline types/epoch.go:61]: why not use *big.Int.Cmp method for comparison? it's unnecessary to marshal number to hex for comparison.

File: types/epoch.go [hunk 7]
```diff
@@ -48,11 +47,24 @@ func (e *Epoch) String() string {
 		return e.name
 	}
 
-	return hexutil.EncodeBig(e.number)
+	return e.number.String()
+}
```
Comment (boqiu) [inline types/epoch.go:51]: Please help to add a util method as following:

`func (e *Epoch) ToInt() (*big.Int, bool)`

**Review by boqiu — CHANGES_REQUESTED** [review]

**Review by wangdayong228 — COMMENTED** [review]
debug is old api, has removed

> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45#-:-MQWKRlFDKvw08pAkUwE:b-ucyoif)* status: 0 of 11 files reviewed, 8 unresolved discussions (waiting on @boqiu)

___
*[interface.go, line 55 at r1](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45#-MQV4ZfS5G5fdOW2amP9-r1-55:-MQVyxPO7iKB3MF7m4HZ:bb7f2mf) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/8e9c91111647bedf3c2cd94fb70a01ee8f6a9256/interface.go#L55)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

rust code has 2nd parameter `epoch`
</blockquote></details>

done
___
*[interface.go, line 89 at r1](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45#-MQV4ZfS5G5fdOW2amP9-r1-89:-MQVzAxfBvaYlvxNbeOd:b-phw4ha) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/8e9c91111647bedf3c2cd94fb70a01ee8f6a9256/interface.go#L89)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

why not use point type, since the default epochs are all defined as pointer types.
</blockquote></details>

because rust api use non-option epoch be parameter
___
*[rpc/metrics.go, line 26 at r1](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45#-MQV4ZfS5G5fdOW2amPB-r1-26:-MQWKKaW2RgQvpypM1i8:bb7f2mf) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/8e9c91111647bedf3c2cd94fb70a01ee8f6a9256/rpc/metrics.go#L26)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

moreover, if go-eth/rpc package also imported, better to ensure go-eth/rpc initialized at first.

So, you can explicitly import eth/rpc package in this file to ensure metrics are initialized correctly.
</blockquote></details>

done
___
*[rpc/metrics.go, line 29 at r1](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45#-MQV4ZfS5G5fdOW2amPB-r1-29:-MQVzffI36U7QesHEsOx:bb7f2mf) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/8e9c91111647bedf3c2cd94fb70a01ee8f6a9256/rpc/metrics.go#L29)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

Why the last one not changed?
</blockquote></details>

done
___
*[types/epoch.go, line 51 at r1](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45#-MQV4ZfS5G5fdOW2amPC-r1-51:-MQWFHNn12atDeNT87Ur:bfxzdkq) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/8e9c91111647bedf3c2cd94fb70a01ee8f6a9256/types/epoch.go#L51)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

Please help to add a util method as following:

`func (e *Epoch) ToInt() (*big.Int, bool)`
</blockquote></details>

done, use Epoch as function receiver to avoid handle nil pointer
___
*[types/epoch.go, line 54 at r1](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45#-MQV4ZfS5G5fdOW2amPC-r1-54:-MQW-Krj69KLkfKSIiJx:bb7f2mf) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/8e9c91111647bedf3c2cd94fb70a01ee8f6a9256/types/epoch.go#L54)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

In golang naming convention, `Equals` is a better name.

Also suggest to rename parameter `input` to `other`
</blockquote></details>

done
___
*[types/epoch.go, line 55 at r1](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45#-MQV4ZfS5G5fdOW2amPC-r1-55:-MQW-U477AtaLQvxLHWq:b-hk3ms6) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/8e9c91111647bedf3c2cd94fb70a01ee8f6a9256/types/epoch.go#L55)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

`e` should not be nil
</blockquote></details>

why? How to do if e is nil?
___
*[types/epoch.go, line 61 at r1](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45#-MQV4ZfS5G5fdOW2amPC-r1-61:-MQW-msI-ahTPqAGZ2E3:bb7f2mf) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/8e9c91111647bedf3c2cd94fb70a01ee8f6a9256/types/epoch.go#L61)):*
<details><summary><i>Previously, boqiu (Bo QIU) wrote…</i></summary><blockquote>

why not use *big.Int.Cmp method for comparison? it's unnecessary to marshal number to hex for comparison.
</blockquote></details>

done

**Review by wangdayong228 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45)* status: 0 of 12 files reviewed, 8 unresolved discussions (waiting on @boqiu)

___
*[types/epoch.go, line 55 at r1](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45#-MQV4ZfS5G5fdOW2amPC-r1-55:-MQWP7sl0ERrstyV79Lj:b-896fix) ([raw file](https://github.com/conflux-chain/go-conflux-sdk/blob/8e9c91111647bedf3c2cd94fb70a01ee8f6a9256/types/epoch.go#L55)):*
<details><summary><i>Previously, wangdayong228 wrote…</i></summary><blockquote>

why? How to do if e is nil?
</blockquote></details>

Done.

**Review by boqiu — APPROVED** [review]
:lgtm:

> Reviewed 7 of 11 files at r1, 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45#-:-MQWVIGxDmSmRaTMabkR:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @wangdayong228)

**Review by boqiu — APPROVED** [review]
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/conflux-chain/go-conflux-sdk/45)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @wangdayong228)

**Pull request merged**